### PR TITLE
Add local cache for video metadata

### LINKED ISSUE #24: Local cache for video info
There should be some local caching going on so we can minimize requests to the youtube api. It should be handled inside `infoextractor.js`

In the youtube api, ETags are used to determine if a resource has changed.
https://developers.google.com/youtube/v3/getting-started#etags

### PR DESCRIPTION
**Pull request by dyc3**
The purpose of this PR is to minimize usage of the youtube API. Closes #24 

Summary:
- Video metadata cache. When video information is requested by calling `getVideoInfo()`, we first check the database to see if it's cached already. If its not there, we retrieve it.
- Partial information retrieval. If certain information is missing from the cache, we can retrieve the bare minimum information needed to fill the gaps.
- `test` environment database config for unit tests to use.
- Unit tests to test caching behavior.

### CODE REVIEW

File: migrations/20190921194802-create-cached-video.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      id: {
+        allowNull: false,
+        autoIncrement: true,
+        primaryKey: true,
+        type: Sequelize.INTEGER,
+      },
+      service: {
+        type: Sequelize.STRING,
+      },
+      service_id: {
```
Comment (adapap) [inline migrations/20190921194802-create-cached-video.js]: camelCase on `service_id` here?

File: infoextract.js [hunk 2]
```diff
@@ -2,13 +2,53 @@ const axios = require("axios");
...
+	getVideoInfo(service, id) {
+		// TODO: check if service is valid
+		// TODO: check if id is valid for service
+		return storage.getVideoInfo(service, id).then(result => {
+			let video = {};
+			for (const key in result) {
+				if (result.hasOwnProperty(key)) {
+					video[key] = result[key];
+				}
+			}
```
Comment (adapap) [inline infoextract.js]: What is the point of checking if `result` has a property `key`? Wouldn't this be better written as `const key in Object.keys(result)`?

Additionally, updating an object with partial properties sounds like a perfect use case for [`Object.assign`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign).

File: infoextract.js [hunk 3]
```diff
@@ -2,13 +2,53 @@ const axios = require("axios");
...
+			let missingInfo = [
+				"title",
+				"description",
+				"thumbnail",
+				"length",
+			].filter(p => !video.hasOwnProperty(p));
```
Comment (adapap) [inline infoextract.js]: Might want to find an alternative to hard-coding this if video properties get changed in the future. Maybe from existing cache data or from database model?

File: infoextract.js (hunk 2 above)
Comment (dyc3) [reply infoextract.js]: The `if (result.hasOwnProperty(key))` came from the vscode template. And yes, `Object.assign` appears to be the better solution.

File: infoextract.js (hunk 3 above)
Comment (dyc3) [reply infoextract.js]: That's a good idea, I'll look into it.

File: migrations/20190921194802-create-cached-video.js (hunk 1 above)
Comment (dyc3) [reply migrations/20190921194802-create-cached-video.js]: Yeah I'll change that

File: storage.js [hunk 4]
```diff
@@ -33,7 +35,82 @@ module.exports = {
...
+			// less than 7 days ago, then the interval is 7 days. The reason for this is that the uploader
+			// is unlikely to change the video info after a week of the original upload. Since we don't store
+			// the upload date, we pretend the original cache date is the upload date. This is potentially an
+			// over optimization.
+			const isCachedInfoValid = lastUpdatedAt.diff(today, "days") <= (origCreatedAt.diff(today, "days") <= 7) ? 7 : 30;
+			let video = {
+				service: cachedVideo.service,
+				id: cachedVideo.serviceId,
+			};
+			// We only invalidate the title and description because those are the only ones that can change.
```
Comment (adapap) [inline storage.js]: The thumbnail can also change. The video author may also unpublish a video; might need to test what the API returns in this case.

File: storage.js [hunk 5]
```diff
@@ -33,7 +35,82 @@ module.exports = {
...
+			}).catch(err => {
+				console.error("Failed to cache video info", err);
+				return false;
+			});
+		});
+	},
+	getVideoInfoFields() {
+		let fields = [];
+		for (let column in CachedVideo.rawAttributes) {
+			if (column === "id" || column === "createdAt" || column === "updatedAt" || column === "serviceId") {
```
Comment (adapap) [inline storage.js]: Just a side-note: for this pattern I usually do something equivalent to `[...].contains(column)`

File: storage.js (hunk 4 above)
Comment (dyc3) [reply storage.js]: The thumbnail may change, but the _url_ to the thumbnail doesn't (needs confirmation).
For example, this video: https://www.youtube.com/watch?v=mOVOS9AjgFs
Has thumbnail: https://img.youtube.com/vi/mOVOS9AjgFs/hqdefault.jpg

File: storage.js (hunk 5 above)
Comment (dyc3) [reply storage.js]: `rawAttributes` is an object

Comment (adapap) [reply storage.js]: In that case you can use `Object.keys(...)` to make the pattern if you want to use the filter method.

**Pull request merged**